plugins/legion-hid/fu-legion-hid-device.c

```diff
@@ -0,0 +1,658 @@
+		    FWUPD_ERROR,
+		    FWUPD_ERROR_BUSY,
+		    "response mismatch filter(%u, %u, %u), retrying...",
+		    main_id,
+		    sub_id,
+		    dev_id);
+	return FALSE;
+}
+
+static gboolean
+fu_legion_hid_device_read_response(FuLegionHidDevice *self,
+				   struct FuStructLegionHidNormalRetryParam *param,
+				   GError **error)
+{
+	return fu_device_retry_full(FU_DEVICE(self),
+				    fu_legion_hid_device_read_normal_response_retry_func,
+				    5,
+				    0,
+				    (gpointer)param,
+				    error);
+}
+
+static gboolean
+fu_legion_hid_device_read_upgrade_response_retry_func(FuDevice *device,
+						      gpointer user_data,
+						      GError **error)
+{
+	struct FuStructLegionHidUpgradeRetryParam *param =
+	    (struct FuStructLegionHidUpgradeRetryParam *)user_data;
+	GByteArray *res = param->res;
+	guint8 main_id = param->main_id;
+	guint8 sub_id = param->sub_id;
+	guint8 dev_id = param->dev_id;
+	guint8 step = param->step;
+	gboolean is_valid_dev_id = FALSE;
+
+	g_byte_array_set_size(res, FU_LEGION_HID_DEVICE_FW_REPORT_LENGTH);
+	if (!fu_udev_device_read(FU_UDEV_DEVICE(device),
+				 res->data,
+				 res->len,
+				 NULL,
+				 FU_LEGION_HID_DEVICE_IO_TIMEOUT,
+				 FU_IO_CHANNEL_FLAG_NONE,
+				 error))
+		return FALSE;
+
+	is_valid_dev_id = (res->data[4] == dev_id) ||
+			  (res->data[4] == 3 && (dev_id == 5 || dev_id == 7)) ||
+			  (res->data[4] == 4 && (dev_id == 6 || dev_id == 8));
+
+	if (res->data[2] == main_id && res->data[3] == sub_id && is_valid_dev_id &&
+	    res->data[7] == step && res->data[9] != FU_LEGION_HID_RESPONSE_STATUS_BUSY)
+		return TRUE;
+
+	g_set_error(error,
+		    FWUPD_ERROR,
+		    FWUPD_ERROR_BUSY,
+		    "response mismatch filter(%u, %u, %u, %u), retrying...",
+		    main_id,
+		    sub_id,
+		    dev_id,
+		    step);
+	return FALSE;
+}
+
+static gboolean
+fu_legion_hid_device_read_upgrade_response(FuLegionHidDevice *self,
+					   struct FuStructLegionHidUpgradeRetryParam *param,
+					   GError **error)
+{
+	return fu_device_retry_full(FU_DEVICE(self),
+				    fu_legion_hid_device_read_upgrade_response_retry_func,
+				    120,
+				    0,
+				    (gpointer)param,
+				    error);
+}
+
+static gboolean
+fu_legion_hid_device_upgrade_start(FuLegionHidDevice *self,
+				   guint8 id,
+				   guint16 crc16,
+				   guint size,
+				   GError **error)
+{
+	g_autoptr(GByteArray) res = NULL;
+	g_autoptr(FuStructLegionHidUpgradeCmd) st_cmd = fu_struct_legion_hid_upgrade_cmd_new();
+	g_autoptr(FuStructLegionHidUpgradeStartParam) st_content =
+	    fu_struct_legion_hid_upgrade_start_param_new();
+	struct FuStructLegionHidUpgradeRetryParam param = {0};
+	guint8 status = 0;
+	fu_struct_legion_hid_upgrade_start_param_set_crc16(st_content, crc16);
+	fu_struct_legion_hid_upgrade_start_param_set_size_high(st_content, (guint8)(size >> 16));
+	fu_struct_legion_hid_upgrade_start_param_set_size_mid(st_content, (guint8)(size >> 8));
+	fu_struct_legion_hid_upgrade_start_param_set_size_low(st_content, (guint8)size);
```

Comment:
i'm guessing size should be defined as a `u24` in the `.rs` file to make this endian safe...